File: ibrdtn/ibrdtn/ibrdtn/data/Bundle.cpp
/*
 * Bundle.cpp
 *
 * Copyright (C) 2011 IBR, TU Braunschweig
 *
 * Written-by: Johannes Morgenroth <morgenroth@ibr.cs.tu-bs.de>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include "ibrdtn/data/Bundle.h"
#include "ibrdtn/data/Serializer.h"
#include "ibrdtn/data/AgeBlock.h"
#include "ibrdtn/data/MetaBundle.h"
#include "ibrdtn/data/BundleID.h"
#include <algorithm>

namespace dtn
{
	namespace data
	{
		Bundle::Bundle(bool zero_timestamp)
		 : PrimaryBlock(zero_timestamp)
		{
			// if the timestamp is not set, add a ageblock
			if (timestamp == 0)
			{
				// add a new ageblock
				push_front<dtn::data::AgeBlock>();
			}
		}

		Bundle::~Bundle()
		{
			clear();
		}

		Bundle::iterator Bundle::begin()
		{
			return _blocks.begin();
		}

		Bundle::iterator Bundle::end()
		{
			return _blocks.end();
		}

		Bundle::const_iterator Bundle::begin() const
		{
			return _blocks.begin();
		}

		Bundle::const_iterator Bundle::end() const
		{
			return _blocks.end();
		}

		bool Bundle::operator==(const BundleID& other) const
		{
			return other == (const PrimaryBlock&)(*this);
		}

		bool Bundle::operator==(const MetaBundle& other) const
		{
			return other == (const PrimaryBlock&)(*this);
		}

		bool Bundle::operator!=(const Bundle& other) const
		{
			return (const PrimaryBlock&)(*this) != (const PrimaryBlock&)other;
		}

		bool Bundle::operator==(const Bundle& other) const
		{
			return (const PrimaryBlock&)(*this) == (const PrimaryBlock&)other;
		}

		bool Bundle::operator<(const Bundle& other) const
		{
			return (const PrimaryBlock&)(*this) < (const PrimaryBlock&)other;
		}

		bool Bundle::operator>(const Bundle& other) const
		{
			return (const PrimaryBlock&)(*this) > (const PrimaryBlock&)other;
		}

		void Bundle::remove(const dtn::data::Block &block)
		{
			for (iterator it = begin(); it != end(); ++it)
			{
				const refcnt_ptr<Block> &b = (*it);
				if (b == &block)
				{
					erase(it);
					return;
				}
			}
		}

		void Bundle::erase(Bundle::iterator b, Bundle::iterator e)
		{
			for (iterator it = b; it != e;)
			{
				_blocks.erase(it++);
			}

			if (size() > 0) {
				// set the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, true);
			}
		}

		void Bundle::erase(iterator it)
		{
			_blocks.erase(it);

			if (size() > 0) {
				// set the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, true);
			}
		}

		void Bundle::clear()
		{
			_blocks.clear();
		}

		dtn::data::PayloadBlock& Bundle::insert(iterator before, ibrcommon::BLOB::Reference &ref)
		{
			if (size() > 0) {
				// remove the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, false);
			}

			dtn::data::PayloadBlock *tmpblock = new dtn::data::PayloadBlock(ref);
			block_elem block( static_cast<dtn::data::Block*>(tmpblock) );

			_blocks.insert(before, block);

			// set the last block bit
			iterator last = end();
			--last;
			(**last).set(dtn::data::Block::LAST_BLOCK, true);

			return (*tmpblock);
		}

		dtn::data::PayloadBlock& Bundle::push_front(ibrcommon::BLOB::Reference &ref)
		{
			dtn::data::PayloadBlock *tmpblock = new dtn::data::PayloadBlock(ref);
			block_elem block( static_cast<dtn::data::Block*>(tmpblock) );
			_blocks.push_front(block);

			// if this was the first element
			if (size() == 1)
			{
				// set the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, true);
			}

			return (*tmpblock);
		}

		dtn::data::PayloadBlock& Bundle::push_back(ibrcommon::BLOB::Reference &ref)
		{
			if (size() > 0) {
				// remove the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, false);
			}

			dtn::data::PayloadBlock *tmpblock = new dtn::data::PayloadBlock(ref);
			block_elem block( static_cast<dtn::data::Block*>(tmpblock) );
			_blocks.push_back(block);

			// set the last block bit
			block->set(dtn::data::Block::LAST_BLOCK, true);

			return (*tmpblock);
		}

		dtn::data::Block& Bundle::push_front(dtn::data::ExtensionBlock::Factory &factory)
		{
			block_elem block( factory.create() );
			_blocks.push_front(block);

			// if this was the first element
			if (size() == 1)
			{
				// set the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, true);
			}

			return (*block);
		}

		dtn::data::Block& Bundle::push_back(dtn::data::ExtensionBlock::Factory &factory)
		{
			if (size() > 0) {
				// remove the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, false);
			}

			block_elem block( factory.create() );
			_blocks.push_back(block);

			// set the last block bit
			block->set(dtn::data::Block::LAST_BLOCK, true);

			return (*block);
		}
		
		Block& Bundle::insert(iterator before, dtn::data::ExtensionBlock::Factory& factory)
		{
			if (size() > 0) {
				// remove the last block bit
				iterator last = end();
				--last;
				(**last).set(dtn::data::Block::LAST_BLOCK, false);
			}

			block_elem block( factory.create() );
			_blocks.insert(before, block);

			// set the last block bit
			iterator last = end();
			--last;
			(**last).set(dtn::data::Block::LAST_BLOCK, true);

			return (*block);
		}

		Size Bundle::size() const
		{
			return _blocks.size();
		}

		bool Bundle::allEIDsInCBHE() const
		{
			if( destination.isCompressable()
			    && source.isCompressable()
			    && reportto.isCompressable()
			    && custodian.isCompressable()
			  )
			{
				for( const_iterator it = begin(); it != end(); ++it ) {
					if( (**it).get( Block::BLOCK_CONTAINS_EIDS ) )
					{
						std::list< EID > blockEIDs = (**it).getEIDList();
						for( std::list< EID >::const_iterator itBlockEID = blockEIDs.begin(); itBlockEID != blockEIDs.end(); ++itBlockEID )
						{
							if( ! itBlockEID->isCompressable() )
							{
								return false;
							}
						}
					}
				}
				return true;
			}
			else
			{
				return false;
			}
		}

		dtn::data::Length Bundle::getPayloadLength() const
		{
			try {
				const dtn::data::PayloadBlock &payload = find<dtn::data::PayloadBlock>();
				return payload.getLength();
			} catch (const NoSuchBlockFoundException&) {
				return 0;
			}
		}

		Bundle::const_iterator Bundle::find(block_t blocktype) const
		{
			return std::find(begin(), end(), blocktype);
		}

		Bundle::iterator Bundle::find(block_t blocktype)
		{
			return std::find(begin(), end(), blocktype);
		}

		Bundle::iterator Bundle::find(const dtn::data::Block &block)
		{
			for (iterator it = begin(); it != end(); ++it)
			{
				if ((&**it) == &block) return it;
			}

			return end();
		}

		Bundle::const_iterator Bundle::find(const dtn::data::Block &block) const
		{
			for (const_iterator it = begin(); it != end(); ++it)
			{
				if ((&**it) == &block) return it;
			}

			return end();
		}

	}
}

File: java/ibrdtnlib/src/ibrdtn/api/object/Bundle.java
/*
 * Bundle.java
 * 
 * Copyright (C) 2011 IBR, TU Braunschweig
 *
 * Written-by: 
 *  Johannes Morgenroth <morgenroth@ibr.cs.tu-bs.de>
 *  Julian Timpner      <timpner@ibr.cs.tu-bs.de>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package ibrdtn.api.object;

import ibrdtn.api.Timestamp;
import java.util.LinkedList;
import java.util.logging.Level;
import java.util.logging.Logger;

public class Bundle {

    private static final Logger logger = Logger.getLogger(Bundle.class.getName());
    protected EID destination;
    protected SingletonEndpoint source;
    protected SingletonEndpoint custodian;
    protected SingletonEndpoint reportto;
    protected long lifetime;
    protected Timestamp timestamp;
    protected Long sequenceNumber;
    protected long procFlags;
    private Long appDataLength = null;
    private Long fragmentOffset = null;
    protected LinkedList<Block> blocks = new LinkedList<Block>();

    public enum Priority {

        BULK,
        NORMAL,
        EXPEDITED
    }

    public enum Flags {

        FRAGMENT(0),
        ADM_RECORD(1),
        NO_FRAGMENT(2),
        CUSTODY_REQUEST(3),
        DESTINATION_IS_SINGLETON(4),
        APP_ACK_REQUEST(5),
        RESERVED_6(6),
        PRIORITY_BIT1(7),
        PRIORITY_BIT2(8),
        CLASSOFSERVICE_9(9),
        CLASSOFSERVICE_10(10),
        CLASSOFSERVICE_11(11),
        CLASSOFSERVICE_12(12),
        CLASSOFSERVICE_13(13),
        RECEPTION_REPORT(14),
        CUSTODY_REPORT(15),
        FORWARD_REPORT(16),
        DELIVERY_REPORT(17),
        DELETION_REPORT(18),
        // DTNSEC FLAGS (these are customized flags and not written down in any draft)
        DTNSEC_REQUEST_SIGN(26),
        DTNSEC_REQUEST_ENCRYPT(27),
        DTNSEC_STATUS_VERIFIED(28),
        DTNSEC_STATUS_CONFIDENTIAL(29),
        DTNSEC_STATUS_AUTHENTICATED(30),
        COMPRESSION_REQUEST(31);
        private int offset;

        Flags(int offset) {
            this.offset = offset;
        }

        public int getOffset() {
            return offset;
        }
    }

    public Bundle() {
    }

    public Bundle(EID destination, long lifetime) {
        setDestination(destination);
        this.lifetime = lifetime;
        setPriority(Priority.NORMAL);
    }

    /**
     * Set a flag in the bundle
     *
     * @param flag a constant from Bundle.Flags
     * @param value true to set the flag, false to clear it
     */
    public void setFlag(Flags flag, boolean value) {
        logger.log(Level.FINE, "Setting {0}", flag.toString());
        if (value) {
            procFlags |= 0b1L << flag.getOffset();
        } else {
            procFlags &= ~(0b1L << flag.getOffset());
        }
    }

    public Boolean getFlag(Flags flag) {
        return (flag.getOffset() & this.procFlags) > 0;
    }

    public Priority getPriority() {
        if (getFlag(Flags.PRIORITY_BIT1)) {
            return Priority.NORMAL;
        }

        if (getFlag(Flags.PRIORITY_BIT2)) {
            return Priority.EXPEDITED;
        }

        return Priority.BULK;
    }

    public void setPriority(Priority p) {
        switch (p) {
            case BULK:
                setFlag(Flags.PRIORITY_BIT1, false);
                setFlag(Flags.PRIORITY_BIT2, false);
                break;

            case EXPEDITED:
                setFlag(Flags.PRIORITY_BIT1, false);
                setFlag(Flags.PRIORITY_BIT2, true);
                break;

            case NORMAL:
                setFlag(Flags.PRIORITY_BIT1, true);
                setFlag(Flags.PRIORITY_BIT2, false);
                break;
        }
    }

    public EID getDestination() {
        return destination;
    }

    public void setDestination(EID destination) {
        this.destination = destination;
        if (destination instanceof SingletonEndpoint) {
            setFlag(Flags.DESTINATION_IS_SINGLETON, true);
        }
    }

    public SingletonEndpoint getSource() {
        return source;
    }

    public void setSource(SingletonEndpoint source) {
        this.source = source;
    }

    public SingletonEndpoint getCustodian() {
        return custodian;
    }

    /**
     * Usually, the custodian is only set by the daemon.
     *
     * @param custodian
     */
    public void setCustodian(SingletonEndpoint custodian) {
        this.custodian = custodian;
    }

    public SingletonEndpoint getReportto() {
        return reportto;
    }

    public void setReportto(SingletonEndpoint reportto) {
        this.reportto = reportto;
    }

    public long getLifetime() {
        return lifetime;
    }

    public void setLifetime(long lifetime) {
        this.lifetime = lifetime;
    }

    public Timestamp getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Timestamp timestamp) {
        this.timestamp = timestamp;
    }

    public Long getSequenceNumber() {
        return sequenceNumber;
    }

    public void setSequenceNumber(Long sequenceNumber) {
        this.sequenceNumber = sequenceNumber;
    }

    public Long getAppDataLength() {
        return appDataLength;
    }

    public void setAppDataLength(Long appDataLength) {
        this.appDataLength = appDataLength;
    }

    public Long getFragmentOffset() {
        return fragmentOffset;
    }

    public void setFragmentOffset(Long fragmentOffset) {
        this.fragmentOffset = fragmentOffset;
    }

    public long getProcFlags() {
        return procFlags;
    }

    public void setProcFlags(long procFlags) {
        this.procFlags = procFlags;
    }

    public boolean isSingleton() {
        int single = (int) (procFlags >> Flags.DESTINATION_IS_SINGLETON.getOffset()) & 0b11;
        if (single == 0) {
            return false;
        } else {
            return true;
        }
    }

    /**
     * Appends a new Block to this bundle
     *
     * @param block the block to append
     */
    public void appendBlock(Block block) {
        if (blocks.size() > 0) {
            blocks.getLast().procflags &= ~(1 << Block.FlagOffset.LAST_BLOCK.ordinal());
        }
        block.procflags |= (1 << Block.FlagOffset.LAST_BLOCK.ordinal());
        blocks.addLast(block);
    }

    public LinkedList<Block> getBlocks() {
        return blocks;
    }

    /**
     * Returns the bundle's payload block, or null if no payload block exists.
     *
     * It it assumed that a maximum of one payload block exists.
     *
     * @return
     */
    public PayloadBlock getPayloadBlock() {
        PayloadBlock payload = null;

        for (Block block : getBlocks()) {
            if (block.getType() == PayloadBlock.type) {
                payload = (PayloadBlock) block;
            }
        }
        return payload;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Bundle: ");
        builder.append(source);
        builder.append(" -> ");
        builder.append(destination);
        builder.append(" ");
        for (Block block : blocks) {
            builder.append("[");
            builder.append(block);
            builder.append("]");
        }
        return builder.toString();
    }
}
